### ISSUE TITLE
Auction Concept

### ISSUE THREAD

**Issue by gingersven-ab** [ArtBlocks/artblocks-contracts#417]
Thanks to pompeyplottin for encouraging me to create this issue.  Feel free to comment / adjust as needed to fit style.

The idea here is to create a minting mechanic that accomplishes the following goals : 
1. Avoids FOMO
2. All minters pay the same price
3. Provide equal opportunity to anyone interested in minting
4. Add monetary component to minting to avoid bot related behavior

The mint process would have up to three distinct phases.  The first would involve a set period of time where any wallets interested in minting can signal their willingness to participate in the mint at a minimum of a pre-determined price (effectively the resting tier of a standard Dutch auction).  In this phase, one wallet is equivalent to one signal.  If the number of wallets signaling a willingness to mint is >= the number of mints in the intended collection the second phase is activated.  In the second phase, there is a bidding period where all wallets have the opportunity to privately bid for a mint.  In the event there are more bids than pieces in the collection, the highest bids matching the collection size are awarded an allow list spot for minting at the lowest bid from that list of highest bids. In the event there are fewer bids than pieces in the collection, the bids are awarded an allow list spot for minting at the lowest bid from all bids.  The remaining pieces in the collection would be available for open minting at the lowest bid from the bid list.

Example : 
- gingersven is releasing a collection of 250 pieces for a minimum of 0.10 ETH
- signaling period for the collection opens at 12p CST on 1/4/2022 and is open for 24 hours
- if after 24 hours there have been 500 signals for minting this collection at a minimum of 0.10 ETH
     - bidding opens at 12p CST on 1/6/2022 and is open for 12 hours
          - 50 bids received for 0.10 ETH, 100 bids received for 0.15 ETH, 100 bids received for 0.18 ETH, 50 bids received for 0.20 ETH, 100 bids received for 0.25 ETH, 50 bids received for 0.50 ETH, 50 bids received for 1.00 ETH
          - the top 250 bids (50 at 1.00E, 50 at 0.50E, 100 at 0.25E, and 50 at 0.20E) are added to an allow list to mint at 0.20 ETH
          - minting is restricted to allow list wallets for one week after which the remaining mints are open to all wallets at 0.20 ETH
- if after 24 hours there have only been 200 signals for minting this collection at a minimum of 0.10 ETH
     - bidding is not necessary, signaling wallets added to an allow list for one week and the remaining mints are open for all wallets at 0.10 ETH

Requirements : 
- signaling is costless
- bidding has some minor cost associated to avoid spam bidding ... perhaps escrow 10% of bid amount until bid period has lapsed after which it is refunded
- bids are not public during the bidding window

**Comment by jakerockland** [comment]
Thank you for filing this @gingersven-ab !!, a couple thoughts on constraints, regarding the requirements bullet points.

> signaling is costless

I do think if you have signaling being costless, it is very very likely to run into an issue of people manipulating / spamming the signaling – I think a financial pledge would be necessary here rather than just a signal.

> bidding has some minor cost associated to avoid spam bidding ... perhaps escrow 10% of bid amount until bid period has lapsed after which it is refunded

Similar to the note above, in order to avoid manipulation here I think it would be important that bidding requires actually locking funds in-minter during the time of the bid.

> bids are not public during the bidding window

This is somewhat possible using a blind-auction type approach (e.g. for a simple example see this blind auction example: https://vyper.readthedocs.io/en/stable/vyper-by-example.html#blind-auction) but it is necessary in this type of approach to pledge an amount that is >= the amount being blind-bid.

**Comment by gingersven-ab** [comment]
Fair point regarding spam of signaling, its a low impact area but figured it'd be frustrating to pay gas each step of the way.  This portion could be eliminate and move straight to blind bidding.  Just adds a mandatory layer of bid period when there may be low enough interest to just sell at the resting price. 

Bidding having a minor cost ... the idea of escrowing a portion of the bid price is intended to avoid the manipulation.  Since this would be an on chain transaction, there's no need to worry about deminimis size spam transactions.  A bigger deal would be manipulating the price higher, which is where the locking some % of the bid size comes into play.  Can't imagine spamming bids of any material size at high bid levels if your ETH is locked up.  

Not sure I follow the reason behind requiring a pledge >= the amount being blind-bid.  The bid wouldn't be executed upon but lead to an allow-list spot at a fixed price at the lowest of the winning bids.  Perhaps its required from a technical perspective because it obfuscates the actual bid cost by setting an arbitrarily high pledge?

In addition to what you mentioned above; my proposal as written has a serious flaw insofar that it has very little penalty for bidding REALLY high to ensure being allocated an allow list spot.  There is an escrow % but that would be refunded at time of auction completion.  Below is an adapted approach to further act as a disincentive for this behavior further... 

- Require 10% of the bid price held in escrow
- Each wallet has a mint price equal to the lowest of the highest # of bids + the square root of the difference between your bid and the lowest of the highest

Example : 
- Minimum winning bid is 0.1 ETH and the highest bid is 10 ETH
- 10 ETH bid would escrow 1 ETH for bidding period and pay 0.4146426545 ETH for their mint
     - 1 ETH = 10% of 10 ETH bid
     - 0.41464 ETH mint price = 0.1 ETH + (sqrt(10 ETH - 0.1 ETH * 100) / 100)

**Comment by justtry1512** [comment]
I am quite excited about this one. best luck guys.
